From 45fcc17b8379c5fe2124a545d1f5e5052c6bf6af Mon Sep 17 00:00:00 2001
From: Youling Tang <tangyouling@kylinos.cn>
Date: Thu, 2 Apr 2026 14:18:32 +0000
Subject: [PATCH v4 06/13] smb/client: check if ntstatus_to_dos_map is sorted

Although the array is sorted at build time, verify the ordering again
when cifs.ko is loaded to avoid potential regressions introduced by
future script changes.

We are going to define 3 functions to check the sort results, introduce the
macro DEFINE_CHECK_SORT_FUNC() to reduce duplicate code.

Signed-off-by: Youling Tang <tangyouling@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/client/cifsfs.c       |  6 ++++++
 fs/smb/client/smb1maperror.c | 32 ++++++++++++++++++++++++++++++++
 fs/smb/client/smb1proto.h    |  1 +
 3 files changed, 39 insertions(+)

diff --git a/fs/smb/client/cifsfs.c b/fs/smb/client/cifsfs.c
index f4bed8d4a072..0541706c36c5 100644
--- a/fs/smb/client/cifsfs.c
+++ b/fs/smb/client/cifsfs.c
@@ -1915,6 +1915,12 @@ init_cifs(void)
 {
 	int rc = 0;
 
+#ifdef CONFIG_CIFS_ALLOW_INSECURE_LEGACY
+	rc = smb1_init_maperror();
+	if (rc)
+		return rc;
+#endif /* CONFIG_CIFS_ALLOW_INSECURE_LEGACY */
+
 	rc = smb2_init_maperror();
 	if (rc)
 		return rc;
diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index fb985d2fc0d9..66ceebbe535e 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -256,3 +256,35 @@ map_and_check_smb_error(struct TCP_Server_Info *server,
 
 	return rc;
 }
+
+#define DEFINE_CHECK_SORT_FUNC(__array, __field)			\
+static int __init __array ## _is_sorted(void)				\
+{									\
+	unsigned int i;							\
+									\
+	/* Check whether the array is sorted in ascending order */	\
+	for (i = 1; i < ARRAY_SIZE(__array); i++) {			\
+		if (__array[i].__field >=				\
+		    __array[i - 1].__field)				\
+			continue;					\
+									\
+		pr_err(#__array " array order is incorrect\n");		\
+		return -EINVAL;						\
+	}								\
+									\
+	return 0;							\
+}
+
+/* ntstatus_to_dos_map_is_sorted */
+DEFINE_CHECK_SORT_FUNC(ntstatus_to_dos_map, ntstatus);
+
+int __init smb1_init_maperror(void)
+{
+	int rc;
+
+	rc = ntstatus_to_dos_map_is_sorted();
+	if (rc)
+		return rc;
+
+	return rc;
+}
diff --git a/fs/smb/client/smb1proto.h b/fs/smb/client/smb1proto.h
index 42569bbcf6fd..dd98d04e837a 100644
--- a/fs/smb/client/smb1proto.h
+++ b/fs/smb/client/smb1proto.h
@@ -234,6 +234,7 @@ int cifs_verify_signature(struct smb_rqst *rqst,
  * smb1maperror.c
  */
 int map_smb_to_linux_error(char *buf, bool logErr);
+int smb1_init_maperror(void);
 int map_and_check_smb_error(struct TCP_Server_Info *server,
 			    struct mid_q_entry *mid, bool logErr);
 
-- 
2.53.0

